okta hosted sign-in widget blog post (#1561)
oktadeveloper/okta-blog: _source/_posts/2025-02-24-okta-hosted-sign-in-widget.md

---
layout: blog_post
title: "From Meh to Wow: Customize Your Okta Sign-In Experience"
author: nesh-popovic
by: internal-contributor
communities: [javascript]

> description: "Tutorial on customizing your Okta Hosted Sign-In Widget to the next level."

Reviewer: ```suggestion
description: "Tutorial on customizing your Okta-hosted Sign-In Widget to the next level."
```

tags: [sign-in-custom-domain, custom-domain, okta-hosted-sign-in]
tweets:
- ""
- ""
- ""
image: blog/okta-hosted-sign-in-widget/sign-in-widget-social-image.jpeg

> type: awareness

Reviewer: ```suggestion
type: conversion
```

---

> The Okta Sign-In Widget is a powerful tool that allows developers to integrate Okta authentication into their web applications seamlessly. But did you know you can also customize the look and feel of the widget to match your application(s)' design, even if the widget is hosted on Okta? Whether you want to modify colors, add custom branding, or adjust the layout, Okta provides flexibility to make the widget fit your brand's identity.

Reviewer: ```suggestion
The Okta Sign-In Widget is a powerful tool that allows developers to integrate Okta authentication into their web applications seamlessly. But did you know you can also customize the look and feel of the widget to match your application's design, even if the widget is hosted on Okta? Whether you want to modify colors, add custom branding, or adjust the layout, Okta provides flexibility to make the widget fit your brand's identity.
```

In this guide, we'll transform the Okta-hosted widget using standard Javascript and CSS and its built-in functionality.

{:toc}

**Before**
{% img blog/okta-hosted-sign-in-widget/okta-hosted-sign-in-widget-before.gif alt:"animated gif of default static Okta hosted sign-in widgets" %}{: .center-image }

Our goal:

**After**
{% img blog/okta-hosted-sign-in-widget/okta-hosted-sign-in-widget-after.gif  alt:"animated gif of customized Okta hosted sign-in widget with sliding window effect for new signups" width:"800" %}{: .center-image }

## Okta Sign-In Widget basics

> Before diving into customization, let's go over a few basics. "Okta-hosted widget" means that when the user needs to authenticate, you redirect them from your application to Okta, and the login screen that the user is presented with is hosted by Okta (it could be on a custom domain). Hence, all the widget code is on Okta itself (and not in the application), which makes it very convenient for your developers.

Reviewer: ```suggestion
Before diving into customization, let's go over a few basics. "Okta-hosted widget" means that when the user needs to authenticate, you redirect them from your application to Okta, and the sign-in screen that the user is presented with is hosted by Okta or on a custom domain. Hence, you can customize the widget and tailor code within Okta's Admin Console, which makes it very convenient for your developers.
```

Thoughts?

**Prerequisites to this tutorial**

> This guide assumes you already have a Single Sign-On (SSO) application that redirects to Okta for authentication. In addition, you already have a customized Okta-hosted widget with a [custom domain](https://help.okta.com/en-us/content/topics/settings/settings-configure-custom-url.htm) enabled to make the necessary edits I will be demonstrating in this tutorial. This tutorial doesn't use the Third-generation widget but uses the Sign-In Widget **version ^7** instead. And lastly, this guide assumes you have [Self-Service Registration (SSR)](https://support.okta.com/help/s/article/sign-up-link-missing-from-okta-login-page-during-sp-initiated-login?language=en_US) also enabled to handle user self-registration.  

Reviewer: ```suggestion
This guide assumes you already have a Single Sign-On (SSO) application that redirects to Okta for authentication. In addition, you already have a customized Okta-hosted widget with a [custom domain](https://help.okta.com/en-us/content/topics/settings/settings-configure-custom-url.htm) enabled to make the necessary edits I will be demonstrating in this tutorial. This tutorial uses the Sign-In Widget **version ^7**. And lastly, this guide assumes you have [Self-Service Registration (SSR)](https://support.okta.com/help/s/article/sign-up-link-missing-from-okta-login-page-during-sp-initiated-login?language=en_US) also enabled to handle user self-registration.  
```

Highly recommend splitting this out into bullet points or something easy to scan for prereqs

If you don't have the above setup just yet, you can test one of our [samples](https://github.com/okta-samples) with a free Developer Edition Account under (Sign up free for Developer Edition)[https://developer.okta.com/signup/]. 

>**Note:** You must use your work email address to sign up for a free developer account.

> Within the samples, please refer to the ones labeled **Okta Hosted Login** or **Redirect Model**. For example, here is one in [React](https://github.com/okta-samples/okta-react-sample) and [Angular](https://github.com/okta-samples/okta-angular-sample). Each sample readme should have directions on creating an SSO application within Okta. Still, for more info, you can always refer to this page on [creating app integrations](https://help.okta.com/en-us/content/topics/apps/apps_app_integration_wizard_oidc.htm.). One last note, be sure you are using the correct authorization server for SSO and that is the [Okta Org Authorization Server](https://developer.okta.com/docs/concepts/auth-servers/#org-authorization-server). 

Reviewer: ```suggestion
Within the samples, please refer to the ones labeled **Okta Hosted Login** or **Redirect Model**. For example, here is one in [React](https://github.com/okta-samples/okta-react-sample) and [Angular](https://github.com/okta-samples/okta-angular-quickstart). Each sample readme should have directions on creating an SSO application within Okta. Still, for more info, you can always refer to this page on [creating app integrations](https://help.okta.com/en-us/content/topics/apps/apps_app_integration_wizard_oidc.htm.). One last note, be sure you are using the correct authorization server for SSO and that is the [Okta Org Authorization Server](https://developer.okta.com/docs/concepts/auth-servers/#org-authorization-server). 
```

For even more examples of how to get started with the Okta-hosted Sign-In Widget, check out this post, which walks you through the basic customizations and setting up a custom domain.

{% excerpt /blog/2023/01/12/signin-custom-domain %}

## Customizing colors and branding on the Okta Sign-In Widget
One of the easiest and basic ways to personalize the Okta Sign-In Widget is through the **Customizations** > **Brands** > **[your custom brand]** > **Theme** tab; here, you can modify the logo, colors, and background. 

## Understanding the Okta Sign-In Widget structure
We're not here to change the button color or simply slap a logo on the widget; we're here to learn how to slice and dice that widget so it dazzles and provides a better user experience. 🙂

We will accomplish this by utilizing standard Javascript and CSS.
If you use the Okta-hosted widget, you will find it under **Customizations** > **Brands** > **[your custom brand]** > **Pages** > **Sign-in** page. Click the **Configure** button and toggle the **Code editor** to **ON**. At this point, you should see some code appear like the one below:

{% raw %}
```html

> <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">

Reviewer: Great. syntax highlighting is working again. Please align indentions. they are all over the place

               // Remove or replace this with your own custom error handler.
               console.log(error.message, error);
           }
       );
</script>
</body>
</html>
```
{% endraw %}

Let's make a note of a few key things:
- `OktaUtil` is a global object that contains all of the widget methods and settings
- `var config = OktaUtil.getSignInWidgetConfig();` gets us all the widget configurations so we can start modifying them 
- `var oktaSignIn = new OktaSignIn(config);` initializes the widget with the config values

## Add custom CSS and third-party libraries to the Okta Sign-In Widget
If you want more fine-grained control over the widget's appearance, Okta allows you to inject your own CSS styles as well as use other libraries (e.g., Bootstrap, jQuery, etc.). You can override specific elements or apply custom styles across the widget. To apply your own CSS, you can either do the inline styling or point to a CSS file (which needs to be accessible from the web). Let's dive in! 

Within the `<head>` section and below the following code,

{% raw %}
```html
<!-- Favicon from theme -->
    <link rel="shortcut icon" href="{{faviconUrl}}" type="image/x-icon"/>

> ```

Reviewer: You need the `endraw` here not after the next code block

insert the following code that points to the Font Awesome library, which is hosted on my AWS S3 bucket (you can use your own Font Awesome or any other library, but you would need to modify the code accordingly).

```css
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css">
```
{% endraw %}

Next, let's add some general styles to achieve the desired look. Just below the following block of code,

{% raw %}
```html
<title>{{pageTitle}}</title>
    {{{SignInWidgetResources}}}

    <style nonce="{{nonceValue}}">
        #login-bg-image-id {
            background-image: {{bgImageUrl}}
        }
    </style>
```
{% endraw %}

add the following code,

> ```css

Reviewer: Alignment is odd.

        top: 0;
        height: 100%;
        width: 50%;
        transform: translateX(0);
        transition: transform 0.6s ease-in-out;
    }
    .overlay-left {
        transform: translateX(-20%);
    }
    .container.right-panel-active .overlay-left {
        transform: translateX(0);
    }
    .overlay-right {
        right: 0;
        transform: translateX(0);
    }
    .container.right-panel-active .overlay-right {
        transform: translateX(20%);
    }
    
    footer {
        background-color: #222;
        color: #fff;
        font-size: 14px;
        bottom: 0;
        position: fixed;
        left: 0;
        right: 0;
        text-align: center;
        z-index: 999;
    }
    footer p {
        margin: 10px 0;
    }
    footer i {
        color: red;
    }
    footer a {
        color: #3c97bf;
        text-decoration: none;
    }
```

Lastly, we will modify the widget CSS to our desired style. I will not explain each here, but it is just a CSS juggling act. Please note that I am not a CSS expert, so the following styles can probably be used more efficiently. 🙂

Still, within the `<head>` section, and just below the code above, insert the following code:

> ```css

Reviewer: Please align indentions and reformat all code snippets to be consistent

        if(clientid == ["CLIENT ID OF YOUR OKTA DASHBOARD APP"]){
            
            $("#reg-overlay").css("display","none"); // hide the registration panel
            
            $("#admin-overlay").css("display","flex"); // display the text if this is an Admin user trying to access the dashboard
            // hide the social buttons container

            $("#okta-login-container div.sign-in-with-idp").css("display","none"); // hide the social buttons
        }

        console.log("Finished manipulating the widget");
    };
</script>
```
{% endraw %}

Feel free to preview your work by clicking the **Save to draft** button and then the **Preview** button at the top, but just know there is one last step before we can see the final result.

## Content Security Policy to allow trusted external resources
The remaining task to do is to customize the default Content Security Policy (CSP) to allow trusted external resources and prevent violations. This will enable us to use external resources, such as the Font Awesome library, Google Fonts, and jQuery. You must add the URLs in the CSP if you use additional libraries or resources.

> >**SECURE BEST PRACTICE**: 

Reviewer: 🙌🏼
